### PR TITLE
Refactor binfmt image.rs

### PR DESCRIPTION
**Pull request by Cardosaum**
Apply pending suggestions from [1](https://github.com/risc0/risc0/pull/1146#pullrequestreview-1743002974) and [2](https://github.com/risc0/risc0/pull/1146#discussion_r1401320191)

### CODE REVIEW

File: risc0/binfmt/src/image.rs [hunk 1]
```diff
@@ -224,25 +217,23 @@ impl MemoryImage {
...
-                    &page[(addr - page_start) as usize..(addr - page_start) as usize + bytes.len()],
-                );
-            }
-        };
+        if let Some(page) = self.pages.get(&page_idx) {
+            bytes.clone_from_slice(
+                &page[(addr - page_start) as usize..(addr - page_start) as usize + bytes.len()],
+            );
+        } else {
+            assert!(
```
Comment (austinabell) [inline risc0/binfmt/src/image.rs:230]: What is the logical difference here if this check is panicked rather than handled as it was before?

Comment (Cardosaum) [reply risc0/binfmt/src/image.rs]: Oh, you mean from `assert!` to `ensure!`? that's actually a good catch!
Thanks!

Comment (Cardosaum) [reply risc0/binfmt/src/image.rs]: We were supposed to remove panics from this code section. But while reverting to the preferred style Frank mentioned I make the mistake of pasting the old version verbatim, whereas I should've replaced the panic for an error variant. Thanks for calling that out!

Comment (austinabell) [reply risc0/binfmt/src/image.rs]: Awesome! Wasn't sure if it was a typo or if it was intentional :)

**Review by flaub — APPROVED** [review]

**Pull request merged**